### PR TITLE
Replace CR and new line with a 0x23CE character

### PR DESCRIPTION
**Pull request by TravisEz13**
# PR Summary

Replace CR and new line with a 0x23CE character

## PR Context

Trying to document how to use logging with Azure Log Analytics and RSyslog converting these to `#012` and `#013`

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs [hunk 1]
```diff
@@ -1525,6 +1525,14 @@ private static bool WriteScriptBlockToLog(ScriptBlock scriptBlock, int segment,
                 // So, replace null characters with the Unicode `SYMBOL FOR NULL`
```
Comment (PoshChan) [inline src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: @TravisEz13, your last commit had 2 failures in `PowerShell-CI-linux`
Basic SysLog tests on Linux.Verifies scriptblock logging
```powershell
Expected regular expression 'Creating Scriptblock text \(1 of 1\):#012\$pid#012& \(\[scriptblock\]::create\("Write\-Verbose 'testheader123' ;Write\-verbose 'after'"\)\)(#012)*ScriptBlock ID: [0-9a-z\-]*#012Path:.*' to match 'Creating Scriptblock text (1 of 1):#012$pid?& ([scriptblock]::create("Write-Verbose 'testheader123' ;Write-verbose 'after'"))?#012#012ScriptBlock ID: e80c2016-0469-4513-9306-fb722676c547#012Path: /tmp/138211ff-582d-4ee2-ac7b-c2f3877b7f93/test01.ps1', but it did not match.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/Logging.Tests.ps1: line 214
214:         $createdEvents[1].Message | Should -Match ($scriptBlockCreatedRegExTemplate -f (Get-RegEx -SimpleMatch $Script.Replace([System.Environment]::NewLine,'#012')))
```
Basic SysLog tests on Linux.Verifies scriptblock logging with null character
```powershell
Expected regular expression 'Creating Scriptblock text \(1 of 1\):#012\$pid#012& \(\[scriptblock\]::create\("Write\-Verbose 'testheader123\$\(\[char\]0x0000\)' ;Write\-verbose 'after'"\)\)(#012)*ScriptBlock ID: [0-9a-z\-]*#012Path:.*' to match 'Creating Scriptblock text (1 of 1):#012$pid?& ([scriptblock]::create("Write-Verbose 'testheader123$([char]0x0000)' ;Write-verbose 'after'"))?#012#012ScriptBlock ID: babe05f7-f77d-45a2-a574-f58c74e7bb46#012Path: /tmp/138211ff-582d-4ee2-ac7b-c2f3877b7f93/test01.ps1', but it did not match.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/Logging.Tests.ps1: line 243
243:         $createdEvents[1].Message | Should -Match ($scriptBlockCreatedRegExTemplate -f (Get-RegEx -SimpleMatch $Script.Replace([System.Environment]::NewLine,'#012')))
```

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs [hunk 2]
```diff
@@ -1525,6 +1525,14 @@ private static bool WriteScriptBlockToLog(ScriptBlock scriptBlock, int segment,
...
                 // We don't just remove the characters to preserve the fact that a null character was there.
                 textToLog = textToLog.Replace('\u0000', '\u2400');
+#if UNIX
+                if (Platform.IsLinux)
+                {
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: This check is not needed.

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs [hunk 3]
```diff
@@ -1525,6 +1525,14 @@ private static bool WriteScriptBlockToLog(ScriptBlock scriptBlock, int segment,
                 // So, replace null characters with the Unicode `SYMBOL FOR NULL`
                 // We don't just remove the characters to preserve the fact that a null character was there.
                 textToLog = textToLog.Replace('\u0000', '\u2400');
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: Do we need to handle `null` the same way for Linux and macOS?

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs [hunk 4]
```diff
@@ -1525,6 +1525,14 @@ private static bool WriteScriptBlockToLog(ScriptBlock scriptBlock, int segment,
...
+#if UNIX
+                if (Platform.IsLinux)
+                {
+                    // Syslog encodes CR and NL to their octal values.
+                    // We will replace them with a unicode charater for easier viewing
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]:
```suggestion
                    // We will replace them with a unicode  'RETURN SYMBOL' (U+23CE) charater for easier viewing
```

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs (hunk 3 above)
Comment (TravisEz13) [reply src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: yes

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs (hunk 2 above)
Comment (TravisEz13) [reply src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: Why?  This is only an issues for SysLog, not OsLog on macOS

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: Got it. Closed

**Comment by daxian-dbw** [comment]
Talked with @TravisEz13 about the concern of creating 3 large strings for `Linux` when the script itself is large.
@TravisEz13 will update the code to address that concern.

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs [hunk 5]
```diff
@@ -1557,6 +1553,52 @@ private static bool WriteScriptBlockToLog(ScriptBlock scriptBlock, int segment,
             return true;
         }
 
+        private static string FormatLogString(string textToLog)
+        {
+            // characters we ar translating from
```
Comment (adityapatwardhan) [inline src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: typo in comment ar -> are

**Review by adityapatwardhan — APPROVED** [review]

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs [hunk 6]
```diff
@@ -1520,11 +1520,7 @@ private static bool WriteScriptBlockToLog(ScriptBlock scriptBlock, int segment,
```
Comment (PoshChan) [inline src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: @TravisEz13, your last commit had 2 failures in `PowerShell-CI-linux`
Basic SysLog tests on Linux.Verifies scriptblock logging
```powershell
Expected regular expression 'Creating Scriptblock text \(1 of 1\):#012\$pid?& \(\[scriptblock\]::create\("Write\-Verbose 'testheader123' ;Write\-verbose 'after'"\)\)(#012)*ScriptBlock ID: [0-9a-z\-]*#012Path:.*' to match 'Creating Scriptblock text (1 of 1):#012$pid?& ([scriptblock]::create("Write-Verbose 'testheader123' ;Write-verbose 'after'"))?#012#012ScriptBlock ID: 75103d64-1b5a-4c80-980b-66f3a0872bc4#012Path: /tmp/c9508df5-4a20-418d-a42a-a04810ee373d/test01.ps1', but it did not match.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/Logging.Tests.ps1: line 214
214:         $createdEvents[1].Message | Should -Match ($scriptBlockCreatedRegExTemplate -f (Get-RegEx -SimpleMatch $Script.Replace([System.Environment]::NewLine,"`u{23CE}")))
```
Basic SysLog tests on Linux.Verifies scriptblock logging with null character
```powershell
Expected regular expression 'Creating Scriptblock text \(1 of 1\):#012\$pid?& \(\[scriptblock\]::create\("Write\-Verbose 'testheader123\$\(\[char\]0x0000\)' ;Write\-verbose 'after'"\)\)(#012)*ScriptBlock ID: [0-9a-z\-]*#012Path:.*' to match 'Creating Scriptblock text (1 of 1):#012$pid?& ([scriptblock]::create("Write-Verbose 'testheader123$([char]0x0000)' ;Write-verbose 'after'"))?#012#012ScriptBlock ID: abdbfad9-56d8-4149-b909-d790e3fcaed1#012Path: /tmp/c9508df5-4a20-418d-a42a-a04810ee373d/test01.ps1', but it did not match.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/Logging.Tests.ps1: line 243
243:         $createdEvents[1].Message | Should -Match ($scriptBlockCreatedRegExTemplate -f (Get-RegEx -SimpleMatch $Script.Replace([System.Environment]::NewLine,"`u{23CE}")))
```

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs [hunk 7]
```diff
@@ -1557,6 +1553,52 @@ private static bool WriteScriptBlockToLog(ScriptBlock scriptBlock, int segment,
             return true;
         }
 
+        private static string FormatLogString(string textToLog)
+        {
+            // characters we are translating from
+
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: Please remove empty line

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs (hunk 6 above)
Comment (PoshChan) [inline src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: @TravisEz13, your last commit had 2 failures in `PowerShell-CI-linux`
Basic SysLog tests on Linux.Verifies scriptblock logging
```powershell
Expected regular expression 'Creating Scriptblock text \(1 of 1\):#012\$pid[?\?]& \(\[scriptblock\]::create\("Write\-Verbose 'testheader123' ;Write\-verbose 'after'"\)\)(#012)*ScriptBlock ID: [0-9a-z\-]*#012Path:.*' to match 'Creating Scriptblock text (1 of 1):#012$pid?& ([scriptblock]::create("Write-Verbose 'testheader123' ;Write-verbose 'after'"))?#012#012ScriptBlock ID: 263aebb2-a3e9-4f37-8121-7c4b992ce77d#012Path: /tmp/9c51aa89-25d4-4e87-a6fe-a717fe13d7fb/test01.ps1', but it did not match.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/Logging.Tests.ps1: line 215
215:         $createdEvents[1].Message | Should -Match ($scriptBlockCreatedRegExTemplate -f (Get-RegEx -SimpleMatch $Script.Replace([System.Environment]::NewLine,"`u{23CE}")))
```
Basic SysLog tests on Linux.Verifies scriptblock logging with null character
```powershell
Expected regular expression 'Creating Scriptblock text \(1 of 1\):#012\$pid[?\?]& \(\[scriptblock\]::create\("Write\-Verbose 'testheader123\$\(\[char\]0x0000\)' ;Write\-verbose 'after'"\)\)(#012)*ScriptBlock ID: [0-9a-z\-]*#012Path:.*' to match 'Creating Scriptblock text (1 of 1):#012$pid?& ([scriptblock]::create("Write-Verbose 'testheader123$([char]0x0000)' ;Write-verbose 'after'"))?#012#012ScriptBlock ID: 59c2e14c-c82f-47ca-9795-f485c2379cb7#012Path: /tmp/9c51aa89-25d4-4e87-a6fe-a717fe13d7fb/test01.ps1', but it did not match.
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/Host/Logging.Tests.ps1: line 244
244:         $createdEvents[1].Message | Should -Match ($scriptBlockCreatedRegExTemplate -f (Get-RegEx -SimpleMatch $Script.Replace([System.Environment]::NewLine,"`u{23CE}")))
```

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs [hunk 8]
```diff
@@ -1557,6 +1553,52 @@ private static bool WriteScriptBlockToLog(ScriptBlock scriptBlock, int segment,
...
+                logBuilder.Replace(CarriageReturnControlChar, ReturnSymbolChar);
+
+                return logBuilder.ToString();
+            }
+            else
+            {
+                return textToLog.Replace(NullControlChar, NullSymbolChar);
+            }
+#else
+                return textToLog.Replace(NullControlChar, NullSymbolChar);
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: Please fix the indentation.

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs (hunk 2 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: In the case it would be good to have a protection comment in the code.

Comment (TravisEz13) [reply src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: I can repeat the comment if you want

Comment (TravisEz13) [reply src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: It is explained on line 1584

File: test/powershell/Host/Logging.Tests.ps1 [hunk 9]
```diff
@@ -130,7 +130,7 @@ function Get-RegEx
     $regex = $regex -replace '\-', '\-'
     $regex = $regex -replace '\$', '\$'
     $regex = $regex -replace '\^', '\^'
-    $regex = $regex -replace "\`u{23CE}", "[`u{23CE}\?]"
+    #$regex = $regex -replace "\`u{23CE}", "[`u{23CE}\?]"
```
Comment (iSazonov) [inline test/powershell/Host/Logging.Tests.ps1]: Can we remove this?

File: test/powershell/Host/Logging.Tests.ps1 [hunk 10]
```diff
@@ -158,7 +158,7 @@ Describe 'Basic SysLog tests on Linux' -Tag @('CI','RequireSudoOnUnix') {
             }
             [string] $powershell = Join-Path -Path $PSHome -ChildPath 'pwsh'
             $scriptBlockCreatedRegExTemplate = @"
-Creating Scriptblock text \(1 of 1\):#012{0}(`u{23CE}|\?|#012)*ScriptBlock ID: [0-9a-z\-]*#012Path:.*
+Creating Scriptblock text \(1 of 1\):#012{0}(⏎|#012)*ScriptBlock ID: [0-9a-z\-]*#012Path:.*
```
Comment (iSazonov) [inline test/powershell/Host/Logging.Tests.ps1]: Why do we need this? `u{23CE} looks more clear for me.

Comment (TravisEz13) [reply test/powershell/Host/Logging.Tests.ps1]: The tests fail if we use the escape sequence.

File: src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs (hunk 6 above)
Comment (PoshChan) [inline src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: @TravisEz13, your last commit had 1 failures in `PowerShell-CI-static-analysis`
Verify Markdown Links.Verify links in /home/vsts/work/1/s/demos/DSC/readme.md.https://docs.microsoft.com/powershell/dsc/overview should work
```powershell
retry of URL failed with error: Response status code does not indicate success: 404 (Not Found).
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 117
117:                                     throw "retry of URL failed with error: $($_.Exception.Message)"

```

Comment (PoshChan) [inline src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs]: @TravisEz13, your last commit had 2 failures in `PowerShell-CI-static-analysis`
Verify Markdown Links.Verify links in /home/vsts/work/1/s/demos/DSC/readme.md.https://docs.microsoft.com/powershell/dsc/overview should work
```powershell
retry of URL failed with error: Response status code does not indicate success: 404 (Not Found).
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 117
117:                                     throw "retry of URL failed with error: $($_.Exception.Message)"

```
Verify Markdown Links.Verify links in /home/vsts/work/1/s/demos/DSC/readme.md.https://docs.microsoft.com/powershell/dsc/overview should work
```powershell
retry of URL failed with error: Response status code does not indicate success: 404 (Not Found).
at <ScriptBlock>, /home/vsts/work/1/s/test/common/markdown/markdown-link.tests.ps1: line 117
117:                                     throw "retry of URL failed with error: $($_.Exception.Message)"

```

File: test/powershell/Host/Logging.Tests.ps1 (hunk 10 above)
Comment (iSazonov) [reply test/powershell/Host/Logging.Tests.ps1]: I tried to copy and paster the code in console and lost the char. Also the tests become unreadable. I wonder why doesn't the Unicode syntax sugar work?

Comment (TravisEz13) [reply test/powershell/Host/Logging.Tests.ps1]: @iSazonov that's going to vary by the font you use.  I worked with @rjmholt to come up with this solution.  The previous solution that was working, was actually just looking for a `?` which is less specific.

Comment (iSazonov) [reply test/powershell/Host/Logging.Tests.ps1]: I remember that we had Unicode chars in files and it was a headache - editors silently break them. After that we made a conclusion to keep all files in ASCII.

**Pull request merged**